src/photon/photon_algorithm.c

```diff
@@ -17,26 +17,27 @@ typedef struct task_queue_entry {
   struct task_queue_entry *next;
 } task_queue_entry;
 
+/** A data structure used to track which objects are available locally and
+ *  which objects are being actively fetched. */
 typedef struct {
   /** Object id of this object. */
   object_id object_id;
-  /** Handle for the uthash table. */
-  UT_hash_handle handle;
-} available_object;
+  /** Handle for the uthash table. NOTE: This handle is used for both the
+   *  scheduling algorithm state's local_objects and fetch_requests tables.
+   *  We must reinforce the uthash invariant that the entry be in either one or
+   *  neither tables. */
+  UT_hash_handle hh;
+} object_entry;
+
+UT_icd task_queue_entry_icd = {sizeof(task_queue_entry *), NULL, NULL, NULL};
 
-/** A data structure used to track which objects are being fetched. */
 typedef struct {
-  /** The object ID that we are trying to fetch. */
+  /** Object id of this object. */
   object_id object_id;
-  /** The local scheduler state. */
-  local_scheduler_state *state;
-  /** The scheduling algorithm state. */
-  scheduling_algorithm_state *algorithm_state;
-  /** The ID for the timer that will time out the current request. */
-  int64_t timer;
-  /** Handle for the uthash table. */
-  UT_hash_handle hh;
-} fetch_object_request;
+  UT_array *dependent_tasks;
+  /** hh for the uthash table. */
+  UT_hash_handle dependency_handle;
```

Comment:
Yeah I mostly called it that so that anyone who uses one of the uthash methods has to explicitly name the right hash handle. Probably, I should change the name of the other handle to describe the hash table also.